src/Session.js

```diff
@@ -36,13 +36,18 @@ export default class Session {
         }
         if (this.state !== Session.State.LOGGING_IN) {
             this.state = Session.State.LOGGING_IN
-            this.sessionTokenPromise = this.loginFunction()
+            try {
+                this.sessionTokenPromise = this.loginFunction().then((tokenObj) => {
+                    this.options.sessionToken = tokenObj.token
+                    this.state = Session.State.LOGGED_IN
+                    return tokenObj.token
+                })
+            } catch (err) {
```

Comment:
This should probably be a `.catch(...)` for the promise, not an actual `try-catch` block, since any errors from `loginFunction` will appear as rejection of the promise, right? 

Before changing this part, please create a test (which fails) first, since I think we're missing coverage on that part. It's a good pattern worth learning anyway: notice bug, create test (should fail!), fix bug, run test (should pass!).

Since there's some non-trivial functionality in `Session.js`, I think it deserves a set of unit tests (in addition to the integration test). There we can easily mock the `_client.loginWith*` functionalities for detailed testing. Plus in the current integration test, the stuff in `describe('Internal state', ...)` would be better off moved to a unit test, allowing you to check that the `loginFunction` is called an expected number of times, as discussed in one of the previous comments.